### PR TITLE
fix: openshift use pods/log to retrieve logs from build

### LINKED ISSUE #2844: oc:build on openshift use pods/log to retrieve logs from build
Hello support,

We running the openshift-maven-plugin against our openshift cluster to build images. The build its self works and pushes to artifactory without any errors.

But we got this message, which shows an error retrieving the logs for that build job. 

```
[INFO] oc: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods ?? is forbidden: User \"system:serviceaccount:ns:user\" cannot get resource \"pods/log\" in API group \"\" in the namespace \"auswamt\"","reason":"Forbidden","details":{"name":"app-s2i-7-build","kind":"pods"},"code":403}
```

The message shows that the plugin need access to the resource of pods/log. If I test this with the oc command (oc logs -f build/app-s2i-9), I can retrieve the logs with a service account without access to pods/log.

The service account is defined with the following rules:

```
rules:
  - verbs:
      - create
      - delete
      - get
      - list
      - patch
      - update
      - watch
    apiGroups:
      - ''
      - build.openshift.io
    resources:
      - buildconfigs
  - verbs:
      - create
    apiGroups:
      - ''
      - build.openshift.io
    resources:
      - buildconfigs/instantiate
      - buildconfigs/instantiatebinary
  - verbs:
      - get
      - list
      - watch
    apiGroups:
      - ''
      - build.openshift.io
    resources:
      - builds
      - builds/log
```

In my opinion, the builds/log resource access should be sufficient to retrive the logs of oc:build. Can you adjust this, as it's not a good idea to give more rights than necessary.

openshift-maven-plugin: 1.16.1

Thanks for your work and best regards,

Steffen

### PR DESCRIPTION
**Pull request by arsenalzp**
## Description

Fix #2844 
This PR  substitutes log retrieval from `pods/log` by `build/logs`.
It removes the necessity for additional rules for building service account.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [X] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] I have read the [contributing guidelines](https://www.eclipse.dev/jkube/contributing)
 - [X] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [X] My code follows the style guidelines of this project
 - [X] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
 - [X] I use [conventional commits](https://www.conventionalcommits.org/) in my commit messages
 - [X] I have performed a self-review of my code
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry
 - [ ] I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly
 - [X] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [X] I have added tests that prove my fix is effective or that my feature works
 - [X] New and existing unit tests pass locally with my changes
 - [] I tested my code in Kubernetes
 - [X] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
### Eclipse JKube [CI Report](https://github.com/eclipse-jkube/ci)

Started new GH workflow run for https://github.com/eclipse-jkube/jkube/pull/3208 (_2024-06-26T19:47:55Z_)

:gear: [JKube E2E Tests (9685826998)](https://github.com/eclipse-jkube/ci/actions/runs/9685826998) :hourglass_flowing_sand:
- :hourglass_flowing_sand: [Bootstrap CI test run (#3208)](https://github.com/eclipse-jkube/ci/actions/runs/9685826998/job/26726834717)

**Comment by arsenalzp** [comment]
Hello colleagues,
Could you please kindly review this PR?

File: jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildServiceIntegrationTest.java [hunk 1]
```diff
@@ -552,6 +562,7 @@ private WebServerEventCollector prepareMockServer(
         .build();
 
     if (!buildConfigExists) {
+      mockServer.expect().get().withPath("/api/v1/namespaces/ns1/pods?labelSelector=openshift.io%2Fbuild.name").andReply(collector.record("build-config-check").andReturn(200, bc)).always();
```
Comment (rohanKanojia) [inline jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildServiceIntegrationTest.java:565]: You've changed the code to use build/logs, but why isn't that reflected in mock expectations? I see this expectation is for listing pods created for build.

Comment (arsenalzp) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildServiceIntegrationTest.java:565]: Yeah, you are right, however tests look for that resource - `pod`. Let me check it again and find a object which is using this API call

Comment (arsenalzp) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildServiceIntegrationTest.java:565]: @rohanKanojia hello
I found the issue: 
https://github.com/fabric8io/kubernetes-client/blob/c8163e53f7cb9344e8f6c85597a7eabe14189bc1/openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/build/BuildOperationsImpl.java#L194

It is looking for `Pod` resource

Comment (rohanKanojia) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildServiceIntegrationTest.java:565]: @arsenalzp : Thanks for clarification!

Comment (arsenalzp) [reply jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildServiceIntegrationTest.java:565]: Could this remark be resolved?

**Comment by rohanKanojia** [comment]
I tested this pull request on CRC with ServiceAccount token configured as described in https://github.com/eclipse-jkube/jkube/issues/2844#issuecomment-2067091501: 
```
CRC version: 2.37.1+36d451
OpenShift version: 4.15.14
```

When a token with restricted permission is used we get this error on v1.16.2 (as reported in :
```
[INFO] --- oc:1.16.2:build (default-cli) @ random-generator ---
[INFO] oc: Using OpenShift build with strategy S2I
[INFO] oc: Running generator spring-boot
[INFO] oc: spring-boot: Using Docker image quay.io/jkube/jkube-java:0.0.23 as base / builder
[INFO] oc: Final output artifact file was not rebuilt since last build. HINT: try to compile and package your application prior to running the container image build task.
[INFO] oc: [random-generator:0.0.1] "spring-boot": Created docker source tar /home/rokumar/work/repos/eclipse-jkube-demo-project/target/docker/random-generator/0.0.1/tmp/docker-build.tar
[INFO] oc: Adding to Secret pullsecret-jkube
[INFO] oc: Using Secret pullsecret-jkube
[INFO] oc: Updating BuildServiceConfig random-generator-s2i for Source strategy
[INFO] oc: Adding to ImageStream random-generator
[INFO] oc: Starting Build random-generator-s2i
[INFO] oc: Waiting for build random-generator-s2i-2 to complete...
[INFO] oc: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods \"random-generator-s2i-2-build\" is forbidden: User \"system:serviceaccount:default:testsa\" cannot get resource \"pods/log\" in API group \"\" in the namespace \"default\"","reason":"Forbidden","details":{"name":"random-generator-s2i-2-build","kind":"pods"},"code":403}
[INFO] oc: Build random-generator-s2i-2 in status Complete
[INFO] oc: Found tag on ImageStream random-generator tag: sha256:112520bef1f2f06c057f4bc69da6925068fd82874c967fead3d8e8eafa84bb71
[INFO] oc: ImageStream random-generator written to /home/rokumar/work/repos/eclipse-jkube-demo-project/target/random-generator-is.yml
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  26.603 s
[INFO] Finished at: 2024-07-29T22:15:26+05:30
[INFO] ------------------------------------------------------------------------
```

However, with these changes I don't see this warning. I'm able to see `oc:build` complete with full logs :heavy_check_mark: 
 :
```
[INFO] --- oc:1.17-SNAPSHOT:build (default-cli) @ random-generator ---
[INFO] oc: Using OpenShift build with strategy S2I
[INFO] oc: Running generator spring-boot
[INFO] oc: spring-boot: Using Docker image quay.io/jkube/jkube-java:0.0.23 as base / builder
[INFO] oc: Final output artifact file was not rebuilt since last build. HINT: try to compile and package your application prior to running the container image build task.
[INFO] oc: [random-generator:0.0.1] "spring-boot": Created docker source tar /home/rokumar/work/repos/eclipse-jkube-demo-project/target/docker/random-generator/0.0.1/tmp/docker-build.tar
[INFO] oc: Creating Secret
[INFO] oc: Creating BuildServiceConfig random-generator-s2i for Source build
[INFO] oc: Creating ImageStream random-generator
[INFO] oc: Starting Build random-generator-s2i
[INFO] oc: Waiting for build random-generator-s2i-1 to complete...
[INFO] oc: Receiving source from STDIN as archive ...
[INFO] oc: time="2024-07-29T16:44:27Z" level=info msg="Not using native diff for overlay, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled"
[INFO] oc: I0729 16:44:27.382362       1 defaults.go:112] Defaulting to storage driver "overlay" with options [mountopt=metacopy=on].
[INFO] oc: Caching blobs under "/var/cache/blobs".
[INFO] oc: Trying to pull quay.io/jkube/jkube-java:0.0.23...
[INFO] oc: Getting image source signatures
[INFO] oc: Copying blob sha256:94ad07648680df54b9e32724cc2febb7f0a01245dfef48e84fa39ce101b056ca
[INFO] oc: Copying blob sha256:24a3fb335248549803dfdfe5f4ea60fcc9e11d5708cc62bb1fb811f608fd094a
[INFO] oc: Copying config sha256:8659a9b5de083556380993b0bbb62cb4907fd85e3c49c6952a216b17e67bdbd9
[INFO] oc: Writing manifest to image destination
[INFO] oc: Generating dockerfile with builder image quay.io/jkube/jkube-java:0.0.23
[INFO] oc: Adding transient rw bind mount for /run/secrets/rhsm
[INFO] oc: STEP 1/9: FROM quay.io/jkube/jkube-java:0.0.23
[INFO] oc: STEP 2/9: LABEL "io.openshift.build.image"="quay.io/jkube/jkube-java:0.0.23"       "io.openshift.build.source-location"="/tmp/build/inputs"       "io.openshift.s2i.destination"="/tmp"
[INFO] oc: STEP 3/9: ENV JAVA_APP_DIR="/deployments"     OPENSHIFT_BUILD_NAME="random-generator-s2i-1"     OPENSHIFT_BUILD_NAMESPACE="default"
[INFO] oc: STEP 4/9: USER root
[INFO] oc: STEP 5/9: COPY upload/src /tmp/src
[INFO] oc: STEP 6/9: RUN chown -R 1000:0 /tmp/src
[INFO] oc: STEP 7/9: USER 1000
[INFO] oc: STEP 8/9: RUN /usr/local/s2i/assemble
[INFO] oc: INFO S2I source build with plain binaries detected
[INFO] oc: INFO S2I binary build from fabric8-maven-plugin detected
[INFO] oc: INFO Copying binaries from /tmp/src/deployments to /deployments ...
[INFO] oc: random-generator-0.0.1.jar
[INFO] oc: INFO Copying deployments from deployments to /deployments...
[INFO] oc: '/tmp/src/deployments/random-generator-0.0.1.jar' -> '/deployments/random-generator-0.0.1.jar'
[INFO] oc: INFO Cleaning up source directory (/tmp/src)
[INFO] oc: STEP 9/9: CMD /usr/local/s2i/run
[INFO] oc: COMMIT temp.builder.openshift.io/default/random-generator-s2i-1:1de12fa8
[INFO] oc: Getting image source signatures
[INFO] oc: Copying blob sha256:064b40dd2dfb098da5b4d43876da5d481047121e1ed560129383b74d30b5fe99
[INFO] oc: Copying blob sha256:3dfb4347f3e4a1e56a0a43a90d3538b63f9924daeb6f1112df242d7729581f40
[INFO] oc: Copying blob sha256:3419a94cf44791b84028944397a58fd0e2ce6f34fc5f5063501bef3dac2553cd
[INFO] oc: Copying config sha256:0d5df84bb47ed9148b9323e3b20659cac2011c6e32d8c1ced43515e2bfa9e331
[INFO] oc: Writing manifest to image destination
[INFO] oc: --> 0d5df84bb47e
[INFO] oc: Successfully tagged temp.builder.openshift.io/default/random-generator-s2i-1:1de12fa8
[INFO] oc: 0d5df84bb47ed9148b9323e3b20659cac2011c6e32d8c1ced43515e2bfa9e331
[INFO] oc: 
[INFO] oc: Pushing image image-registry.openshift-image-registry.svc:5000/default/random-generator:0.0.1 ...
[INFO] oc: Getting image source signatures
[INFO] oc: Copying blob sha256:3419a94cf44791b84028944397a58fd0e2ce6f34fc5f5063501bef3dac2553cd
[INFO] oc: Copying blob sha256:94ad07648680df54b9e32724cc2febb7f0a01245dfef48e84fa39ce101b056ca
[INFO] oc: Copying blob sha256:24a3fb335248549803dfdfe5f4ea60fcc9e11d5708cc62bb1fb811f608fd094a
[INFO] oc: Copying config sha256:0d5df84bb47ed9148b9323e3b20659cac2011c6e32d8c1ced43515e2bfa9e331
[INFO] oc: Writing manifest to image destination
[INFO] oc: Successfully pushed image-registry.openshift-image-registry.svc:5000/default/random-generator@sha256:3b9072562b2cc94d476f81238ee38189ea8347d074959559f47d089b68ef5bc6
[INFO] oc: Push successful
[INFO] oc: Build random-generator-s2i-1 in status Complete
[INFO] oc: Found tag on ImageStream random-generator tag: sha256:3b9072562b2cc94d476f81238ee38189ea8347d074959559f47d089b68ef5bc6
[INFO] oc: ImageStream random-generator written to /home/rokumar/work/repos/eclipse-jkube-demo-project/target/random-generator-is.yml
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  50.382 s
[INFO] Finished at: 2024-07-29T22:14:52+05:30
[INFO] ------------------------------------------------------------------------
```

File: jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildServiceIntegrationTest.java [hunk 2]
```diff
@@ -126,8 +132,12 @@ void init(@TempDir Path temporaryFolder) throws Exception {
     jKubeBuildTarArchiver = mockConstruction(JKubeBuildTarArchiver.class, (mock, ctx) ->
         when(mock.createArchive(any(File.class), any(BuildDirs.class), eq(ArchiveCompression.none)))
         .thenReturn(emptyDockerBuildTar));
-
-    resourceConfig = ResourceConfig.builder().namespace("ns1").build();
+    
+    props = new Properties();
+    props.setProperty("buildName", projectName);
```
Comment (rohanKanojia) [inline jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/openshift/OpenshiftBuildServiceIntegrationTest.java:137]: Is this really required? Even if I remove it test passes. 

Is it a leftover from when you were facing test failures https://github.com/eclipse-jkube/jkube/issues/2844#issuecomment-2075857092 ?

**Comment by rohanKanojia** [comment]
@arsenalzp : Looks good to me :+1: (Just a minor comment related to a leftover property)

**Comment by arsenalzp** [comment]
> @arsenalzp : Looks good to me 👍 (Just a minor comment related to a leftover property)

Thank you!
It was pleased to work on this project.

**Comment by rohanKanojia** [comment]
@arsenalzp : :+1: . Could you please resolve minor conflicts?

**Comment by arsenalzp** [comment]
> @arsenalzp : 👍 . Could you please resolve minor conflicts?

Hello,
I gonna do it today.

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**